### PR TITLE
use the new sdk batch method

### PR DESCRIPTION
**Pull request by joewagner**
Note: This is ready for review, but the tests will be failing until an SDK with the batch upgrades is published. That should happen Apr 25 2023.

## Overview
Now that the Registry contract and the SDK support batching writes and creates into single transactions, clients can start using this feature.  The CLI is one of those clients.   This PR allows the cli user to take advantage of batching via the `write` and `create` command.

## Details
The statement string being passed to the associated command is now being parsed and if it contains SQL that need to use the new batching feature, then the statement is converted into an all-or-nothing single transaction batch.  If the statement does **not** need to be batched, then the more simple (less gas) single table transaction is sent.

It's worth noting that if the CLI user does not want an all-or-nothing transaction, then they will need to send each statement as a separate cli command.  This might be the case if someone is trying to insert a large amount of data and they don't want everything to be reverted if one of the statements has an error.

### CODE REVIEW

File: src/commands/create.ts [hunk 1]
```diff
@@ -74,13 +74,50 @@ export const handler = async (argv: Arguments<Options>): Promise<void> => {
...
+      .replace(/\n/g, "")
+      .replace(/\r/g, "")
+      .replace(/\s+/g, " ")
+      .trim();
+
+    // Parse the statement to see if more than one table is affected.
+    // If yes, then combine into separate statements and batch.
+    // If no, then process via single statement.
+    const statements = statement.split(";").filter((stmt) => stmt.trim());
+    const normalized = await Promise.all(
```
Comment (joewagner) [inline src/commands/create.ts:94]: We have to do some parsing in the CLI before we normalize because the go sqlparser will error if you try to pass a single string with multiple create queries.

File: src/commands/write.ts [hunk 2]
```diff
@@ -55,14 +55,78 @@ export const handler = async (argv: Arguments<Options>): Promise<void> => {
...
+      return;
+    }
+
+    // TODO: taking an arbitrary sql string and returning each individual statement with the associated tableId
+    //    seems very useful in general.  We should expose this via the SDK or parser.
+    const statementsById = (
+      await Promise.all(
+        normalized.statements.map(async function (stmt) {
+          // re-normalize so we can be sure we've isolated each statement and it's tableId
+          const norm = await normalize(stmt);
```
Comment (joewagner) [inline src/commands/write.ts:96]: This seems like a common pattern.  We want to take a sql statement string and create an Array of Objects where each Object has a tableId and statement string.
Maybe we can expose this as a helper in the SDK or the wasm parser?

File: test/create.test.ts [hunk 3]
```diff
@@ -28,7 +31,7 @@ describe("commands/create", function () {
   });
 
   test("throws if chain not provided", async function () {
-    const [account] = getAccounts();
+    const [account] = accounts;
```
Comment (joewagner) [inline test/create.test.ts:34]: Each time `getAccounts()` is called we create 25 new Wallet instances, we might as well save memory here.

File: test/create.test.ts [hunk 4]
```diff
@@ -81,16 +84,34 @@ describe("commands/create", function () {
...
+    equal(res, "error parsing statement: syntax error at position 32 near ')'");
+  });
+
+  test("throws when mixing create and write statements", async function () {
+    const [account] = accounts;
+    const privateKey = account.privateKey.slice(2);
+    const consoleError = spy(console, "error");
+    await yargs([
+      "create",
+      "create table fooz (a int);insert into fooz (a) values (1);",
```
Comment (joewagner) [inline test/create.test.ts:98]: Added some tests for incorrect usage of batching

File: test/create.test.ts [hunk 5]
```diff
@@ -203,8 +224,44 @@ CREATE TABLE cooltable (invalid)
...
   });
 
+  test("passes with two create statements", async function () {
+    const [account] = accounts;
+    const privateKey = account.privateKey.slice(2);
+    const consoleLog = spy(console, "log");
+    await yargs([
+      "create",
+      `create table first_table (id int primary key, name text);
+      create table second_table (id int primary key, name text);`,
```
Comment (joewagner) [inline test/create.test.ts:234]: This is the new batch create functionality

File: test/create.test.ts [hunk 6]
```diff
@@ -221,26 +278,21 @@ CREATE TABLE cooltable (invalid)
...
-          transactionHash.startsWith("0x")
-        );
-      }, "does not match")
-    );
+
+    const res = consoleLog.getCall(0).firstArg;
+    const value = JSON.parse(res);
+    const { prefix, name, chainId, tableId, transactionHash } = value.meta.txn;
+
+    equal(prefix, "file_test");
```
Comment (joewagner) [inline test/create.test.ts:286]: I was having trouble with the usage of `assert.calledWith(` because no matter what went wrong with the test the error message was "does not match" and there was no line number.  In some cases console.log had been called twice making things even harder to track down.
I switched to only looking at the first call and using separate assertions for each property.  If this seems like a good pattern, we could make a separate PR with changes to the rest of the tests.

File: test/setup.ts [hunk 7]
```diff
@@ -9,6 +9,9 @@ if (!globalThis.fetch) {
   (globalThis as any).Response = Response;
 }
 
+// TODO: most tests rely on a spy on the global `console.log`. This means we must
+//    use silent: true here.  As an alternative we could explore using a `logger`
+//    that can be mocked, or spied on, or expose an extension api depending on the test.
 const lt = new LocalTableland({ silent: true });
```
Comment (joewagner) [inline test/setup.ts:15]: Just added a note here.  
If we create a "logger" abstraction, we could enable the tests to spy on the "logger" instead of the global console object.  That would allow us to see the network logs during tests.
If that seems useful, it could be included in another PR.

File: test/create.test.ts (hunk 6 above)
Comment (awmuncy) [reply test/create.test.ts:322]: I'm good with this, it's easier to read.

File: test/setup.ts (hunk 7 above)
Comment (awmuncy) [reply test/setup.ts:15]: I'm okay with adding a logger object, though I'm not sure I'd prioritize it, I'd leave that judgement up to you.

**Review by awmuncy — APPROVED** [review]
Looks good!

**Comment by joewagner** [comment]
@awmuncy I had to rebase this, so it might need another 👍

**Review by awmuncy — APPROVED** [review]
